// gatsby-source-datocms@5.1.5/hooks/sourceNodes/createTypes/utils/getBase64.js
"use strict";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]); if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var fs = require('fs/promises');

var path = require('path');

var resizeUrl = require('./resizeUrl');

var queryString = require('query-string');

var _require = require('gatsby-core-utils'),
    fetchRemoteFile = _require.fetchRemoteFile;

var magicNumbersToMimeType = {
  ffd8ffe0: 'image/jpeg',
  '89504e47': 'image/png',
  '47494638': 'image/gif',
  '52494646': 'image/webp'
};

function download(_x, _x2) {
  return _download.apply(this, arguments);
}

function _download() {
  _download = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(url, cache) {
    var absolutePath, fileBody, base64, magicNumber, extension, mimeType;
    return regeneratorRuntime.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return fetchRemoteFile({
              url: url,
              cache: cache
            });

          case 2:
            absolutePath = _context2.sent;
            _context2.next = 5;
            return fs.readFile(absolutePath);

          case 5:
            fileBody = _context2.sent;
            base64 = fileBody.toString("base64");
            magicNumber = fileBody.toString('hex', 0, 4);
            extension = path.extname(absolutePath).split('.').pop();
            mimeType = magicNumbersToMimeType[magicNumber] || "image/".concat(extension);
            return _context2.abrupt("return", "data:".concat(mimeType, ";base64,").concat(base64));

          case 11:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _download.apply(this, arguments);
}

module.exports = /*#__PURE__*/function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref, cache) {
    var forceBlurhash, format, src, width, height, _src$split, _src$split2, baseUrl, query, _url, result, imgixParams, url, _result;

    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            forceBlurhash = _ref.forceBlurhash, format = _ref.format, src = _ref.src, width = _ref.width, height = _ref.height;
            _src$split = src.split('?'), _src$split2 = _slicedToArray(_src$split, 2), baseUrl = _src$split2[0], query = _src$split2[1];

            if (!(!baseUrl.startsWith('https://www.datocms-assets.com/') || format === 'png' && !forceBlurhash)) {
              _context.next = 15;
              break;
            }

            _url = resizeUrl({
              url: src,
              width: width,
              height: height
            }, 20);
            _context.prev = 4;
            _context.next = 7;
            return download(_url, cache);

          case 7:
            result = _context.sent;
            return _context.abrupt("return", result);

          case 11:
            _context.prev = 11;
            _context.t0 = _context["catch"](4);
            console.log("Error downloading ".concat(_url, " to generate blurred placeholder!: ").concat(_context.t0.message));
            return _context.abrupt("return", null);

          case 15:
            imgixParams = queryString.parse(query);
            imgixParams.lqip = 'blurhash';
            url = "".concat(baseUrl, "?").concat(queryString.stringify(imgixParams));
            _context.prev = 18;
            _context.next = 21;
            return download(url, cache);

          case 21:
            _result = _context.sent;
            return _context.abrupt("return", _result);

          case 25:
            _context.prev = 25;
            _context.t1 = _context["catch"](18);
            console.log("Error downloading ".concat(url, " to generate Blurhash placeholder!: ").concat(_context.t1.message));
            return _context.abrupt("return", null);

          case 29:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[4, 11], [18, 25]]);
  }));

  return function (_x3, _x4) {
    return _ref2.apply(this, arguments);
  };
}();
