// gatsby-source-datocms@5.1.5/hooks/sourceNodes/createTypes/site/DatoCmsGlobalSeo.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _require = require('datocms-client'),
    camelizeKeys = _require.camelizeKeys;

var resolveUsingAttribute = function resolveUsingAttribute(key, definition) {
  return _objectSpread({}, definition, {
    resolve: function resolve(node) {
      return node[key];
    }
  });
};

module.exports = function (_ref) {
  var actions = _ref.actions,
      schema = _ref.schema,
      generateType = _ref.generateType;
  actions.createTypes([schema.buildObjectType({
    name: generateType('GlobalSeo'),
    extensions: {
      infer: false
    },
    fields: {
      siteName: resolveUsingAttribute('site_name', {
        type: 'String'
      }),
      titleSuffix: resolveUsingAttribute('title_suffix', {
        type: 'String'
      }),
      twitterAccount: resolveUsingAttribute('twitter_account', {
        type: 'String'
      }),
      facebookPageUrl: resolveUsingAttribute('facebook_page_url', {
        type: 'String'
      }),
      fallbackSeo: {
        type: generateType('SeoField'),
        resolve: function resolve(node) {
          var result = camelizeKeys(node.fallback_seo);
          return result;
        }
      }
    }
  })]);
};
