// gatsby-source-datocms@5.1.5/hooks/createSchemaCustomization/index.js
"use strict";

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _require = require('humps'),
    pascalize = _require.pascalize;

var createTypes = require('../sourceNodes/createTypes');

var _require2 = require('../../errorMap'),
    prefixId = _require2.prefixId,
    CODES = _require2.CODES;

var CascadedContext = require('../../cascadedContext');

var _require3 = require('../../utils'),
    getLoader = _require3.getLoader;

module.exports = /*#__PURE__*/function () {
  var _ref3 = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref, _ref2) {
    var actions, getNode, getNodesByType, reporter, parentSpan, schema, store, cache, apiToken, previewMode, environment, apiUrl, instancePrefix, pageSize, logApiCalls, statePerQuery, errorText, loader, context, activity;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            actions = _ref.actions, getNode = _ref.getNode, getNodesByType = _ref.getNodesByType, reporter = _ref.reporter, parentSpan = _ref.parentSpan, schema = _ref.schema, store = _ref.store, cache = _ref.cache;
            apiToken = _ref2.apiToken, previewMode = _ref2.previewMode, environment = _ref2.environment, apiUrl = _ref2.apiUrl, instancePrefix = _ref2.instancePrefix, pageSize = _ref2.pageSize, logApiCalls = _ref2.logApiCalls;
            statePerQuery = new WeakMap();
            actions.createResolverContext({
              getQueryContext: function getQueryContext(key) {
                var queryState = statePerQuery.get(key);

                if (!queryState) {
                  queryState = {
                    localeState: new CascadedContext({
                      reporter: reporter
                    }),
                    fallbackLocalesState: new CascadedContext({
                      reporter: reporter
                    })
                  };
                  statePerQuery.set(key, queryState);
                }

                return queryState;
              }
            });

            if (!apiToken) {
              errorText = "API token must be provided!";
              reporter.panic({
                id: prefixId(CODES.MissingAPIToken),
                context: {
                  sourceMessage: errorText
                }
              }, new Error(errorText));
            }

            _context.next = 7;
            return getLoader({
              cache: cache,
              apiToken: apiToken,
              previewMode: previewMode,
              environment: environment,
              apiUrl: apiUrl,
              pageSize: pageSize,
              logApiCalls: logApiCalls,
              loadStateFromCache: !!process.env.GATSBY_WORKER_ID
            });

          case 7:
            loader = _context.sent;
            context = {
              entitiesRepo: loader.entitiesRepo,
              actions: actions,
              getNode: getNode,
              getNodesByType: getNodesByType,
              schema: schema,
              store: store,
              cache: cache,
              generateType: function generateType(type) {
                return "DatoCms".concat(instancePrefix ? pascalize(instancePrefix) : '').concat(type);
              }
            };

            if (process.env.GATSBY_WORKER_ID) {
              _context.next = 17;
              break;
            }

            activity = reporter.activityTimer("loading DatoCMS schema", {
              parentSpan: parentSpan
            });
            activity.start();
            _context.next = 14;
            return loader.load();

          case 14:
            _context.next = 16;
            return loader.saveStateToCache(cache);

          case 16:
            activity.end();

          case 17:
            createTypes(context);

          case 18:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function (_x, _x2) {
    return _ref3.apply(this, arguments);
  };
}();
